### DISCUSSION TITLE
How to cleanly kill and restart XWayland?

### DISCUSSION THREAD

**Discussion by edburns** [microsoft/wslg#597] [Q&A]
# Executive Summary

Due to #15 I am in need of a way to kill and restart XWayland cleanly. Please tell me how.

# Details

Hello WSLg community,

I was highly motivated to move from WSL with VcXsrv on Windows 10 to WSLg on Windows 11 by this extremely friction creating bug: [Clipboard stops working in Windows client](https://sourceforge.net/p/vcxsrv/bugs/114/).  Imagine my disappointment when I realize the same kind of bug exists in WSLg (#15).  

My remediation when using WSL on Windows 10 with VcXsrv was to simply kill and restart VcXsrv when the copy/paste stopped working. Looking at your [README](https://github.com/microsoft/wslg#wslgd) I am not clear on the best way to do the same remediation here. Is there some way to kill and restart XWayland?

Thanks,

Ed

**Comment by Biswa96** [comment]
It is somewhat possible. But one has to manually setup the WSLGd process from ground. This is literally what I am doing with WSLg in Windows 10. The summary goes like this - start user distribution -> start system distribution -> set all the environment variables which `init` does (e.g. WAYLAND_DISPLAY, DISPLAY, XDG_RUNTIME_DIR etc.) -> run WSLGd in system distribution.

**Reply by edburns** [reply]
Thanks for your reply @Biswa96. Is it possible to enter the system distribution and kill and restart WSLGd?

**Reply by Biswa96** [reply]
`wsl.exe --system` (officially)

**Reply by edburns** [reply]
That seems to take me into the user distribution, because when I do ps -a, I see my user processes.

**Reply by Biswa96** [reply]
Do you see /usr/bin/WSLGd process?

**Reply by edburns** [reply]
Now I do.  I had to do `ps -elf`.

**Reply by edburns** [reply]
So, when the problem happens, can I simply do `wsl --system`, kill the `WSLGd` process and restart  `/usr/bin/WSLGd`?

**Reply by Biswa96** [reply]
That should be the way. I just provided the hint about it. I am not using Windows 11 now (probably never will).

**Reply by edburns** [reply]
This absolutely did not work. When I started `/usr/bin/WSLGd` in the system distribution, an apparently infinite list of processes spawned and I had to force shut down the Windows 11 host machine. Still looking for a solution.

**Comment by onomatopellan** [comment]
On latest WSL from the store you can restart WSLg with **Left-Ctrl + Left-Alt + Backspace**.

To enable it, in PowerShell run `notepad.exe $Env:USERPROFILE\.wslgconfig`

and save the file with this content:
```
[system-distro-env]
WESTON_RDPRAIL_SHELL_ALLOW_ZAP=true
```
After `wsl.exe --shutdown` you should be able to restart WSLg manually.
